File: 684-redundant-connection/redundant-connection.java
class Solution {
    private int[] parent;

    private int find(int node) {
        if (parent[node] == node) {
            return node;
        }
        return parent[node] = find(parent[node]); 
    }

    private boolean union(int u, int v) {
        int rootU = find(u);
        int rootV = find(v);
        if (rootU == rootV) return false;
        parent[rootU] = rootV;
        return true;
    }

    public int[] findRedundantConnection(int[][] edges) {
        int n = edges.length;
        parent = new int[n + 1];
        for (int i = 1; i <= n; i++) {
            parent[i] = i;
        }

        for (int[] edge : edges) {
            if (!union(edge[0], edge[1])) {
                return edge;
            }
        }

        return new int[0];
    }
}

File: 684-redundant-connection/redundant-connection.py
class Solution:

  def findRedundantConnection(self, edges: list[list[int]]) -> list[int]:
    n = len(edges)
    parent = list(range(n + 1))

    def find(node: int) -> int:
      if parent[node] != node:
        parent[node] = find(parent[node])  # Path compression
      return parent[node]

    def union(u: int, v: int) -> bool:
      root_u = find(u)
      root_v = find(v)
      if root_u == root_v:
        return False
      parent[root_u] = root_v
      return True

    for u, v in edges:
      if not union(u, v):
        return [u, v]

    return []